// vite-plugin-vue-server-ref/dist/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  __markAsModule(target);
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toModule = (module2) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};

// src/index.ts
__export(exports, {
  default: () => VitePluginServerRef
});

// src/constant.ts
var URL_PREFIX = ["/@server-ref/", "/@server-reactive/"];
var VIRTUAL_PREFIX = ["server-ref:", "server-reactive:"];
var WS_EVENT = "vue-server-ref";
var PREFIXES = [...URL_PREFIX, ...VIRTUAL_PREFIX];

// src/utils.ts
var import_ufo = __toModule(require("ufo"));

// src/diff/apply.ts
function apply(obj, diff) {
  if (typeof obj !== typeof diff)
    return diff;
  if (typeof diff === "object") {
    if (Array.isArray(diff))
      return diff;
    let out;
    if (Array.isArray(obj)) {
      out = obj;
      for (const i in diff)
        out[i] = apply(obj[i], diff[i]);
    } else {
      out = obj;
      for (const i in diff)
        out[i] = apply(obj[i], diff[i]);
    }
    return out;
  }
  return diff;
}

// src/client.ts
var UNDEFINED = "__UNDEFINED___";
function parse(json) {
  return JSON.parse(json, (key, value) => value === UNDEFINED ? void 0 : value);
}

// src/utils.ts
function getBodyJson(req) {
  return new Promise((resolve, reject) => {
    let body = "";
    req.on("data", (chunk) => body += chunk);
    req.on("error", reject);
    req.on("end", () => {
      try {
        resolve(parse(body) || {});
      } catch (e) {
        reject(e);
      }
    });
  });
}
function get(obj, key) {
  return key.split(".").reduce((acc, key2) => acc == null ? void 0 : acc[key2], obj);
}
function set(obj, key, value) {
  const keys = key.split(".");
  let acc = obj;
  keys.slice(0, -1).forEach((key2) => {
    if (!acc[key2])
      acc[key2] = {};
    acc = acc[key2];
  });
  acc[keys[keys.length - 1]] = value;
}
function parseId(id) {
  for (const pre of PREFIXES) {
    if (id.startsWith(pre)) {
      const { pathname: key, search } = (0, import_ufo.parseURL)(id.substr(pre.length));
      const query = (0, import_ufo.parseQuery)(search);
      return {
        key: key.replace(/\//g, "."),
        type: pre.includes("ref") ? "ref" : "reactive",
        prefix: pre,
        diff: "diff" in query
      };
    }
  }
}

// src/generate.ts
function genCode(options, id) {
  var _a;
  const { state, defaultValue, clientVue, debug, debounce } = options;
  const { key, type, prefix, diff } = id;
  const access = type === "ref" ? "data.value" : "data";
  return `
import { ${type}, watch } from "${clientVue}"
import { randId, stringify, parse, define, apply,${type === "reactive" ? " reactiveSet," : ""}${diff ? " clone, diff" : ""} } from "vite-plugin-vue-server-ref/client"

const data = ${type}(${JSON.stringify((_a = get(state, key)) != null ? _a : defaultValue(key))})

let onSet = []
let onPatch = []

define(data, '$syncUp', true)
define(data, '$syncDown', true)
define(data, '$paused', false)
define(data, '$onSet', fn => onSet.push(fn))
define(data, '$onPatch', fn => onPatch.push(fn))

if (import.meta.hot) {
  const id = randId()
  let skipWatch = false
  let timer = null
  ${diff ? "let copy = null" : ""}

  function post(payload) {
    ${debug ? `console.log("[server-ref] [${key}] outgoing", payload)` : ""}
    return fetch('${prefix + key}', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: stringify(payload)
    })
  }

  ${diff ? `
  function makeClone() {
    copy = clone(${access})
  }

  function getDiff() {
    return diff(${access}, copy)
  }
  ` : ""}

  function applyPatch(patch, trigger = true) {
    skipWatch = true
    if (trigger) {
      onPatch.forEach(fn => fn(patch))
      ${debug ? `console.log("[server-ref] [${key}] patch incoming", patch)` : ""}
    }
    apply(${access}, patch)
    ${diff ? "makeClone()" : ""}
    skipWatch = false
  }

  function applySet(newData) {
    skipWatch = true
    onSet.forEach(fn => fn(newData))
    ${type === "ref" ? "data.value = payload.data" : "reactiveSet(data, newData)"}
    ${debug ? `console.log("[server-ref] [${key}] incoming", newData)` : ""}
    ${diff ? "makeClone()" : ""}
    skipWatch = false
  }
 
  ${debug ? `console.log("[server-ref] [${key}] ref", data)` : ""}
  ${debug ? `console.log("[server-ref] [${key}] initial", ${access})` : ""}
 
  import.meta.hot.on("${WS_EVENT}", (payload) => {
    if (!data.$syncDown || data.$paused || payload.key !== "${key}" || payload.source === id)
      return
    if (payload.patch)
      applyPatch(payload.patch)
    else
      applySet(payload.data)
  })

  define(data, '$patch', async (patch) => {
    if (!data.$syncUp || data.$paused)
      return false
    applyPatch(patch, false)
    return post({
      source: id,
      patch,
      timestamp: Date.now(),
    })
  })

  watch(data, (v) => {
    if (timer)
      clearTimeout(timer)
    if (!data.$syncUp || data.$paused || skipWatch)
      return

    timer = setTimeout(()=>{
      post({
        source: id,
        ${diff ? "patch: getDiff()" : `data: ${access}`},
        timestamp: Date.now(),
      })
    }, ${debounce})
  }, { flush: 'sync', deep: true })
}
else {
  define(data, '$patch', async () => false)
}

export default data
`.replace(/\n\s*\n/gm, "\n");
}

// src/options.ts
var defaultOptions = {
  state: {},
  debounce: 10,
  debug: false,
  defaultValue: () => void 0,
  clientVue: "vue",
  onChanged: () => {
  }
};
function resolveOptions(options) {
  return Object.assign({}, defaultOptions, options);
}

// src/index.ts
function VitePluginServerRef(options = {}) {
  const resolved = resolveOptions(options);
  const { state } = resolved;
  const idMaps = {};
  return {
    name: "vite-plugin-vue-server-ref",
    resolveId(id) {
      const idx = VIRTUAL_PREFIX.findIndex((pre) => id.startsWith(pre));
      if (idx > -1)
        return URL_PREFIX[idx] + id.slice(VIRTUAL_PREFIX[idx].length);
      return URL_PREFIX.some((pre) => id.startsWith(pre)) ? id : null;
    },
    configureServer(server) {
      server.middlewares.use(async (req, res, next) => {
        var _a;
        if (!req.url || req.method !== "POST")
          return next();
        const id = parseId(req.url);
        if (!id)
          return;
        const key = id.key;
        const payload = await getBodyJson(req);
        for (const id2 of idMaps[key] || []) {
          const module2 = server.moduleGraph.getModuleById(id2);
          if (module2)
            server.moduleGraph.invalidateModule(module2);
        }
        if (payload.patch)
          set(state, key, apply(get(state, key), payload.patch));
        else
          set(state, key, payload.data);
        server.ws.send({
          type: "custom",
          event: WS_EVENT,
          data: __spreadProps(__spreadValues({}, payload), { key })
        });
        (_a = options.onChanged) == null ? void 0 : _a.call(options, key, get(state, key), payload.patch, payload.timestamp);
        res.write("");
        res.end();
      });
    },
    load(id) {
      const res = parseId(id);
      if (!res)
        return;
      if (!idMaps[res.key])
        idMaps[res.key] = new Set();
      idMaps[res.key].add(id);
      return genCode(resolved, res);
    }
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {});
